# NBA Database: common_player_info

Downloaded from Kaggle on March 13, 2023.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import data sets
path = r'C:\Users\felix\Documents\PSU\class\NBA'

player = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'common_player_info.csv'), index_col = False)

### Basic Statistics

In [3]:
player.head()

person_id first_name    last_name  display_first_last  \
0         51    Mahmoud   Abdul-Rauf  Mahmoud Abdul-Rauf   
1       1505      Tariq  Abdul-Wahad   Tariq Abdul-Wahad   
2      76007       John    Abramovic      John Abramovic   
3     203112     Quincy          Acy          Quincy Acy   
4     200801     Hassan        Adams        Hassan Adams   

  display_last_comma_first display_fi_last         player_slug  \
0      Abdul-Rauf, Mahmoud   M. Abdul-Rauf  mahmoud-abdul-rauf   
1       Abdul-Wahad, Tariq  T. Abdul-Wahad   tariq-abdul-wahad   
2          Abramovic, John    J. Abramovic      john-abramovic   
3              Acy, Quincy          Q. Acy          quincy-acy   
4            Adams, Hassan        H. Adams        hassan-adams   

             birthdate           school country  ...  \
0  1969-03-09 00:00:00  Louisiana State     USA  ...   
1  1974-11-03 00:00:00   San Jose State  France  ...   
2  1919-02-09 00:00:00            Salem     USA  ...   
3  1990-10-06 00:00:00           Baylor     USA  ...   
4  1984-06-20 00:00:00          Arizona     USA  ...   

                 playercode from_year  to_year  dleague_flag nba_flag  \
0        mahmoud_abdul-rauf    1990.0   2000.0             N        Y   
1         tariq_abdul-wahad    1997.0   2003.0             N        Y   
2  HISTADD_brooms_abramovic    1946.0   1947.0             N        Y   
3                quincy_acy    2012.0   2018.0             Y        Y   
4              hassan_adams    2006.0   2008.0             N        Y   

  games_played_flag draft_year draft_round  draft_number greatest_75_flag  
0                 Y       1990           1             3                N  
1                 Y       1997           1            11                N  
2                 Y  Undrafted   Undrafted     Undrafted                N  
3                 Y       2012           2            37                N  
4                 Y       2006           2            54                N  

[5 rows x 33 columns]

In [4]:
player.shape

(2457, 33)

In [5]:
player.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2457 entries, 0 to 2456
Data columns (total 33 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   person_id                         2457 non-null   int64  
 1   first_name                        2457 non-null   object 
 2   last_name                         2457 non-null   object 
 3   display_first_last                2457 non-null   object 
 4   display_last_comma_first          2457 non-null   object 
 5   display_fi_last                   2457 non-null   object 
 6   player_slug                       2457 non-null   object 
 7   birthdate                         2457 non-null   object 
 8   school                            2453 non-null   object 
 9   country                           2457 non-null   object 
 10  last_affiliation                  2457 non-null   object 
 11  height                            2399 non-null   object 
 12  weight

In [6]:
player.describe()

person_id       weight   season_exp       team_id    from_year  \
count  2.457000e+03  2396.000000  2457.000000  2.457000e+03  2445.000000   
mean   3.389107e+05   211.014190     5.120879  1.330706e+09  1989.496933   
std    5.837598e+05    26.498226     4.509179  6.104306e+08    22.570921   
min    3.000000e+00   133.000000     0.000000  0.000000e+00  1946.000000   
25%    7.621100e+04   190.000000     2.000000  1.610613e+09  1974.000000   
50%    7.787600e+04   210.000000     3.000000  1.610613e+09  1992.000000   
75%    2.023220e+05   230.000000     8.000000  1.610613e+09  2009.000000   
max    1.631466e+06   325.000000    23.000000  1.610613e+09  2022.000000   

           to_year  
count  2445.000000  
mean   1993.884254  
std      23.139779  
min    1946.000000  
25%    1978.000000  
50%    1998.000000  
75%    2015.000000  
max    2022.000000

### Data Cleaning

In [7]:
# reformatting column id to str

player['person_id'] = player['person_id'].astype('str')

In [8]:
# checking for missing values

player.isnull().sum()

person_id                             0
first_name                            0
last_name                             0
display_first_last                    0
display_last_comma_first              0
display_fi_last                       0
player_slug                           0
birthdate                             0
school                                4
country                               0
last_affiliation                      0
height                               58
weight                               61
season_exp                            0
jersey                              583
position                             40
rosterstatus                          0
games_played_current_season_flag      0
team_id                               0
team_name                           427
team_abbreviation                   427
team_code                           427
team_city                           427
playercode                            0
from_year                            12


In [9]:
# checking for duplicate rows

df_dups = player[player.duplicated()]

In [10]:
df_dups

Empty DataFrame
Columns: [person_id, first_name, last_name, display_first_last, display_last_comma_first, display_fi_last, player_slug, birthdate, school, country, last_affiliation, height, weight, season_exp, jersey, position, rosterstatus, games_played_current_season_flag, team_id, team_name, team_abbreviation, team_code, team_city, playercode, from_year, to_year, dleague_flag, nba_flag, games_played_flag, draft_year, draft_round, draft_number, greatest_75_flag]
Index: []

[0 rows x 33 columns]

In [11]:
# checking for mixed data types in every column

for col in player.columns.tolist():
    weird = (player[[col]].applymap(type) != player[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (player[weird]) > 0:
        print(col)

school
height
jersey
position
team_name
team_abbreviation
team_code
team_city
draft_round
draft_number


These coincide with our columns that are missing values.

### Data Wrangling

1. Height
2. Weight
3. School
4. from year, to year

#### Height

Converting height from feet-inch format to inches, float64 data type

Imputing missing values with mean 78.051271

In [12]:
# making 2 new lists, one to hold ft value, one to hold in value
# if value is missing, setting to 0

feet = []
inch = []

for ht in player['height']:
    if pd.isna(ht) == True:
        feet.append('0')
        inch.append('0')
    else:
        x = ht.split('-')
        feet.append(x[0])
        inch.append(x[1])

In [13]:
# converting data type from str to int

feet = [eval(i) for i in feet]
inch = [eval(j) for j in inch]

In [14]:
len(feet)

2457

In [15]:
# new list to hold total inch height

ftin = []

for i in range(2457):
    ftin.append((feet[i] * 12) + inch[i])

In [16]:
# setting missing values back to nan

for i in range(2457):
    if ftin[i] == 0:
        ftin[i] = float('nan')

In [17]:
player['height_inches'] = ftin

In [18]:
player['height_inches'].describe()

count    2399.000000
mean       78.051271
std         3.579357
min        66.000000
25%        75.000000
50%        78.000000
75%        81.000000
max        91.000000
Name: height_inches, dtype: float64

In [19]:
# imputing missing height values

player['height_inches'].fillna(78.051271, inplace=True)

#### Weight

Imputing missing values with mean 211.014190

In [20]:
player['weight'].describe()

count    2396.000000
mean      211.014190
std        26.498226
min       133.000000
25%       190.000000
50%       210.000000
75%       230.000000
max       325.000000
Name: weight, dtype: float64

In [21]:
# imputing missing weight values

player['weight'].fillna(211.014190, inplace=True)

#### School

There are only 4 rows missing school values, so we can go look up the missing values and fill them in through Excel.

In [22]:
player[player['school'].isnull() == True]

person_id first_name  last_name display_first_last  \
1106   1630211      Karim       Mane         Karim Mane   
1970    203141      Mirza  Teletovic    Mirza Teletovic   
1989   1630271    Brodric     Thomas     Brodric Thomas   
2088      2757       Beno      Udrih         Beno Udrih   

     display_last_comma_first display_fi_last      player_slug  \
1106              Mane, Karim         K. Mane       karim-mane   
1970         Teletovic, Mirza    M. Teletovic  mirza-teletovic   
1989          Thomas, Brodric       B. Thomas   brodric-thomas   
2088              Udrih, Beno        B. Udrih       beno-udrih   

                birthdate school                 country  ... from_year  \
1106  2000-05-16 00:00:00    NaN                 Senegal  ...    2020.0   
1970  1985-09-17 00:00:00    NaN  Bosnia and Herzegovina  ...    2012.0   
1989  1997-01-28 00:00:00    NaN                     USA  ...    2020.0   
2088  1982-07-05 00:00:00    NaN                Slovenia  ...    2004.0   

     to_year  dleague_flag  nba_flag games_played_flag draft_year draft_round  \
1106  2020.0             Y         Y                 Y  Undrafted   Undrafted   
1970  2017.0             N         Y                 Y  Undrafted   Undrafted   
1989  2021.0             Y         Y                 Y  Undrafted   Undrafted   
2088  2016.0             N         Y                 Y       2004           1   

     draft_number  greatest_75_flag height_inches  
1106    Undrafted                 N          75.0  
1970    Undrafted                 N          81.0  
1989    Undrafted                 N          77.0  
2088           28                 N          75.0  

[4 rows x 34 columns]

#### from_year, to_year

There are only 12 missing values in these rows (they pertain to newer players), so we can search up their start years and fill them in through Excel.

In [23]:
player[player['from_year'].isnull() == True]

person_id first_name last_name  display_first_last  \
84     1631262      Jules   Bernard       Jules Bernard   
216    1630622      Jalen  Crutcher      Jalen Crutcher   
379     203583      Abdul     Gaddy         Abdul Gaddy   
470    1630564    RaiQuan      Gray        RaiQuan Gray   
697    1630562       Matt      Hurt           Matt Hurt   
742    1631300      Drake  Jeffries      Drake Jeffries   
909    1631452       Noah  Kirkwood       Noah Kirkwood   
1024   1631171     Justin     Lewis        Justin Lewis   
1099   1630588      Makur     Maker         Makur Maker   
1245   1630627     Isaiah    Miller       Isaiah Miller   
1479   1631342    Daeqwon   Plowden     Daeqwon Plowden   
2322   1631466  Jeenathan  Williams  Jeenathan Williams   

     display_last_comma_first display_fi_last         player_slug  \
84             Bernard, Jules      J. Bernard       jules-bernard   
216           Crutcher, Jalen     J. Crutcher      jalen-crutcher   
379              Gaddy, Abdul        A. Gaddy         abdul-gaddy   
470             Gray, RaiQuan         R. Gray        raiquan-gray   
697                Hurt, Matt         M. Hurt           matt-hurt   
742           Jeffries, Drake     D. Jeffries      drake-jeffries   
909            Kirkwood, Noah     N. Kirkwood       noah-kirkwood   
1024            Lewis, Justin        J. Lewis        justin-lewis   
1099             Maker, Makur        M. Maker         makur-maker   
1245           Miller, Isaiah       I. Miller       isaiah-miller   
1479         Plowden, Daeqwon      D. Plowden     daeqwon-plowden   
2322      Williams, Jeenathan     J. Williams  jeenathan-williams   

                birthdate                     school      country  ...  \
84    2000-01-21 00:00:00                       UCLA          USA  ...   
216   1999-07-18 00:00:00                     Dayton          USA  ...   
379   1992-01-26 00:00:00                 Washington          USA  ...   
470   1999-07-07 00:00:00              Florida State          USA  ...   
697   2000-04-20 00:00:00                       Duke          USA  ...   
742   1999-01-22 00:00:00                    Wyoming          USA  ...   
909   1999-12-27 00:00:00                    Harvard       Canada  ...   
1024  2002-04-12 00:00:00                  Marquette          USA  ...   
1099  2000-11-04 00:00:00                     Howard  South Sudan  ...   
1245  1997-11-09 00:00:00  North Carolina-Greensboro          USA  ...   
1479  1998-08-29 00:00:00              Bowling Green          USA  ...   
2322  1999-02-12 00:00:00                    Buffalo          USA  ...   

     from_year to_year  dleague_flag  nba_flag games_played_flag draft_year  \
84         NaN     NaN             Y         N                 Y  Undrafted   
216        NaN     NaN             Y         N                 Y  Undrafted   
379        NaN     NaN             Y         N                 Y       2013   
470        NaN     NaN             Y         N                 Y       2021   
697        NaN     NaN             Y         N                 N  Undrafted   
742        NaN     NaN             Y         N                 N  Undrafted   
909        NaN     NaN             Y         N                 N  Undrafted   
1024       NaN     NaN             N         N                 N  Undrafted   
1099       NaN     NaN             Y         N                 N  Undrafted   
1245       NaN     NaN             Y         N                 Y  Undrafted   
1479       NaN     NaN             Y         N                 Y  Undrafted   
2322       NaN     NaN             Y         N                 N  Undrafted   

     draft_round draft_number  greatest_75_flag height_inches  
84     Undrafted    Undrafted                 N     78.051271  
216    Undrafted    Undrafted                 N     78.051271  
379            0            0                 N     78.051271  
470            2           59                 N     78.051271  
697    Undrafted    Undrafte

### Saving Temporary Work

In [24]:
player.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'common_player_info_clean.csv'))

### Excel Cleaning

- Removed index column
- Removed column height for repeating values, moved height_inches over in its place
- Reformatted birthdate column to Short Date format (removes the hours that was there before)

Column school ended up having 34 rows missing data, so I used the info from column last_affiliation to fill in the missing values. I can do this because the column last_affiliation is made up of a combination of columns school and country.

- Imputed missing school values from last_affiliation values that we have.
- Removed column last_affiliation now for repeating values

Column position has 40 rows missing data, so I looked up each player's position from nba.com and filled in the missing information. I also learned about the different positions in basketball and that each category has a purpose even if some of them sound the same. This only resolved 15 rows though, so the rest of the rows I had to look up and reference different sources (espn.com, basketball-reference.com, gleague.nba.com) to get the rest of the missing positions.

https://www.nba.com/players

https://jr.nba.com/basketball-positions/

- Entered missing position labels manually from data off official nba website
- Entered missing position labels manually from looking up their positions off basketball-related websites
- Removed column jersey due to column having messy data, in addition to being a redundant id variable for each player/team

Columns from_year and to_year both have 12 rows missing data (all the same), so I looked up each of the players to find out their start and end years. Most of the players are new to the competitive scene and entered the NBA draft just this year/last year, so I found out their draft years for the from_year column. Since they're all still active, the to_year column has the value 2022, which is the value for this NBA season.

- Entered missing from_year, to_year values manually from looking up their draft years off basketball-related websites

There are 97 rows missing data in the draft_round column and 126 rows missing data in the draft_number column. Using basketball-reference.com, I found the players draft years and pick order, so I manually entered those numbers in. Due to the NBA draft changing format in 1966 though, the beginning years have a different format compared to more recent years, so it's uncertain whether I'll end up using these columns.

https://www.basketball-reference.com/draft/

- Entered missing draft_round, draft_number values manually from data off basketball-reference.com
- Values missing from basketball-reference draft are labeled as Undrafted
- In columns draft_round, draft_number changed values Undrafted into 0

Columns display_last_comma_first, display_fi_last, player_slug just say the same information as column display_first_last, just in different formats. Therefore I removed these columns to reduce the amount of space the data set takes up. Column playercode also seemes to reference the same information, and the formatting widely varies between the players' name and having HIST_ADD_ included in front. I ended up removing this column as well.

- Removed columns display_last_comma_first, display_fi_last, player_slug for repeated information
- Removed column playercode for repeated information, inconsistent formatting

Columns team_id, team_name, team_abbreviation, and team_code all refer to the same kind of information, wtih team_city referencing where each team resides (which is usually included in the team name as well). However, there are 427 rows missing this information in the dataset, too many to go through and fill in. Additionally, players almost always play for more than one team, so the team a player last played for may not be the team that fans associate them with (ex. Michael Jordan last played for the Washington Wizards, but he is most well known for his time on the Chicago Bulls). Additionally, it is unclear whether the team listed is for the player's current team or the team that drafted them. Michael Jordan has the Chicago Bulls listed for him (which is the team that drafted him, not the team that he finished his career with), but Anthony Davis has the Los Angeles Lakers listed for him (which is the team that he currently plays for, not the team that drafted him (New Orleans Hornets/Pelicans)). Due to the volume of missing data, as well as the uncertainty behind the meaning of these columns, I decided to remove them from my wrangled data set.

- Removed columns team_id, team_name, team_abbreviation, team_code, team_city

### Data Profiling (Excel)

Count (Qualitative Variables): Used the Advanced Sort & Filter function in the data tab to filter only unique values.

Min Function: =MIN(A2:A2458)

Max Function: =MAX(A2:A2458)

Mean Function: =AVERAGE(A2:A2458)

Mode Function: =MODE.MULT(A2:A2458)

Mode Function (Qualitative Variables): =INDEX(A2:A2458,MODE.MULT(MATCH(A2:A2458,A2:A2458,0)))

### Wrangled Data Basic Statistics

In [25]:
wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'common_player_info_wrangled.csv'), index_col = False)

In [26]:
wrangled.head()

person_id first_name    last_name  display_first_last  birthdate  \
0         51    Mahmoud   Abdul-Rauf  Mahmoud Abdul-Rauf   3/9/1969   
1       1505      Tariq  Abdul-Wahad   Tariq Abdul-Wahad  11/3/1974   
2      76007       John    Abramovic      John Abramovic   2/9/1919   
3     203112     Quincy          Acy          Quincy Acy  10/6/1990   
4     200801     Hassan        Adams        Hassan Adams  6/20/1984   

            school country  height_inches  weight  season_exp  ...  \
0  Louisiana State     USA           73.0   162.0           9  ...   
1   San Jose State  France           78.0   235.0           7  ...   
2            Salem     USA           75.0   195.0           2  ...   
3           Baylor     USA           79.0   240.0           7  ...   
4          Arizona     USA           76.0   220.0           2  ...   

  games_played_current_season_flag from_year to_year  dleague_flag  nba_flag  \
0                                N      1990    2000             N         Y   
1                                N      1997    2003             N         Y   
2                                N      1946    1947             N         Y   
3                                N      2012    2018             Y         Y   
4                                N      2006    2008             N         Y   

  games_played_flag draft_year draft_round draft_number greatest_75_flag  
0                 Y       1990           1            3                N  
1                 Y       1997           1           11                N  
2                 Y  Undrafted   Undrafted    Undrafted                N  
3                 Y       2012           2           37                N  
4                 Y       2006           2           54                N  

[5 rows x 22 columns]

In [27]:
wrangled.shape

(2457, 22)

In [28]:
wrangled.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2457 entries, 0 to 2456
Data columns (total 22 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   person_id                         2457 non-null   int64  
 1   first_name                        2457 non-null   object 
 2   last_name                         2457 non-null   object 
 3   display_first_last                2457 non-null   object 
 4   birthdate                         2457 non-null   object 
 5   school                            2457 non-null   object 
 6   country                           2457 non-null   object 
 7   height_inches                     2457 non-null   float64
 8   weight                            2457 non-null   float64
 9   season_exp                        2457 non-null   int64  
 10  position                          2457 non-null   object 
 11  rosterstatus                      2457 non-null   object 
 12  games_

In [29]:
wrangled.describe()

person_id  height_inches       weight   season_exp    from_year  \
count  2.457000e+03    2457.000000  2457.000000  2457.000000  2457.000000   
mean   3.389107e+05      78.051271   211.014190     5.120879  1989.650387   
std    5.837598e+05       3.536840    26.167087     4.509179    22.622690   
min    3.000000e+00      66.000000   133.000000     0.000000  1946.000000   
25%    7.621100e+04      75.000000   190.000000     2.000000  1974.000000   
50%    7.787600e+04      78.000000   210.000000     3.000000  1992.000000   
75%    2.023220e+05      81.000000   229.000000     8.000000  2010.000000   
max    1.631466e+06      91.000000   325.000000    23.000000  2022.000000   

           to_year  
count  2457.000000  
mean   1994.021571  
std      23.166283  
min    1946.000000  
25%    1978.000000  
50%    1998.000000  
75%    2015.000000  
max    2022.000000

In [30]:
# checking for missing values

wrangled.isnull().sum()

person_id                           0
first_name                          0
last_name                           0
display_first_last                  0
birthdate                           0
school                              0
country                             0
height_inches                       0
weight                              0
season_exp                          0
position                            0
rosterstatus                        0
games_played_current_season_flag    0
from_year                           0
to_year                             0
dleague_flag                        0
nba_flag                            0
games_played_flag                   0
draft_year                          0
draft_round                         0
draft_number                        0
greatest_75_flag                    0
dtype: int64

In [31]:
# checking for mixed data types in every column

for col in wrangled.columns.tolist():
    weird = (wrangled[[col]].applymap(type) != wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (wrangled[weird]) > 0:
        print(col)